### ISSUE TITLE
Configurable branch options for Github Integration

### ISSUE THREAD

**Issue by SamHutchings** [getsentry/sentry#55763]
### Problem Statement

Hi,

I recently ran into an problem where a Sentry issue had been resolve, started reoccuring, but the issue remained resolved and was not marked as a regression. The exact scenario is detailed below:

- Issue `[FOO]` was created in Sentry.
- A developer pushed commit `comm_1` to a branch with `FIXES [FOO]` into the commit description.
- The Sentry Github Integration marked the issue `[FOO]` as resolved in `comm_1`
- A PR was then opened which contained `comm_1` with `FIXES [FOO]` in the PR description.
- The PR was merged using the squash strategy, therefore creating a new commit `comm_2` in the main branch.
- The Sentry Github Integration marked the issue `[FOO]` as resolved in `comm_2`
- As `comm_1` was never part of a release, the issue was never marked as a regression, even though `comm_2` had been released and the issue had recurred.

On some further testing, it seemed that any commit pushed to _any_ branch with `FIXES [FOO]` will immediately mark the Sentry issue `[FOO]` as resolved, whether the commit is later merged into main or left in a branch forever. 

To me this seems like potentially dangerous behaviour - if a developer opens a branch and pushes what they think is a fix, then forget about it and go do something else, there's no evidence to anyone else that the Sentry issue may be ongoing, or that is isn't actually resolved.

In the particular case noted above, it was doubly misleading as `comm_2` _was_ part of a release, and the issue reoccurred after that release. But because Sentry was still waiting for `comm_1` to be released, it never got marked as a regression. 

I did open a support ticket about this, but apparently the behaviour is expected, which is why i've opened this as a feature request. 

### Solution Brainstorm

There are a couple of ways I can think of to mitigate this risk. 

1. Have the branch(es) that the GitHub integration looks at be configurable in the integration settings. In our particular case, we only want to mark an issue as resolved if the commit is merged into `main`. I'm sure there are other workflows that do want to look at branches other than `main`/`master`, but still may not want to look at all other branches, so a configurable branch pattern or similar would be very useful.

2. Have later commit fixes supersede earlier commit fixes - if the issue is resolved in `comm_1`, then `comm_2`, and then `comm_2` is released and the problem reoccurs, this seems like a regression even if `comm_1` was never released. This seems more like a bug to me, although it's entirely possible I'm not thinking of a scenario in which it would be useful. 

### Product Area

Settings - Integrations

**Comment by Dhrumil-Sentry** [comment]
@brianthi this seems related to improvements being made to the issue resolution flows

**Comment by adrian-delgado** [comment]
> it seemed that any commit pushed to any branch with FIXES [FOO] will immediately mark the Sentry issue [FOO] as resolved, whether the commit is later merged into main or left in a branch forever.

I noticed the same thing happening with the `Bitbucket Integration`

**Comment by lobsterkatie** [comment]
> I noticed the same thing happening with the `Bitbucket Integration`

Yeah, that makes sense, as this is a matter of how our resolution logic works, regardless of the source of the commit data.

I agree that the current behavior is misleading at best. I've added this to our backlog and will bring it up with our PMs and update here once I have more information about when it might be roadmapped.

**Comment by morgankrey** [comment]
This is wild that this is an unmodifiable default. Please, please take a look at fixing this.

**Comment by minori-fh** [comment]
relevant ticket: https://sentry.zendesk.com/agent/tickets/178012

**Comment by MaximeLaurenty** [comment]
To add to this:
I'm not sure "configurable branch" solves all use cases. I personnally like having a "resolve in release X" once a PR is open and waiting for review and not see the issue anymore.
The issue though is that if the PR is never merged (or merged in a feature branch that gets then merged in master), the release never happens and indeed we are blind on that issue.
-> Having a setting to limit how long an issue can be in a "resolved in release X" without release X being merged could solve this? This setting would need to be adjustable by project to cater for everyone's workflows I guess. With this, whatever happens to the PR, issue will appear again as unresolved after some time if it happens again.

**Comment by lobsterkatie** [comment]
This is good feedback, thanks @MaximeLaurenty. We'll update here as we have a better idea when we might take this on.

**Comment by nicolaslegrand91** [comment]
This has also been a problem for me. 
The commit that allegedly fixed an issue was squashed when merging the PR, and the problem re-appeared a few weeks later, but we did not notice it until a client complained.

@SamHutchings 's first option would work for me, and I could suggest an even simpler fix in the meantime: add an option to **completely disable issue resolution based on commit message.** 

This has been problematic enough for us to add:
- a precommit that blocks any commit containing `Fixes SLUG-ID` for all our projects slugs, or any similar message (based on [this regex](https://github.com/getsentry/sentry/blob/f7f2acbf35dcb69a6965270adf186e4bee01b9a2/src/sentry/utils/groupreference.py#L12-L22))
- a CI check that fails if such a pattern appears in a PR description or title.

**Comment by MaximeLaurenty** [comment]
Any insights where we can access this setting as it seems completed?
